File: 05diff/92problem/SquareDigitChains.java
import java.util.ArrayList;
public class SquareDigitChains {
    public static void main(String[] commandLine) {
        long start = System.currentTimeMillis();
        int one = 0;
        int eightynine = 0;
        for(int i = 1; i <10000001; i++) {
            if (createChain(i) == 1) one++;
            else eightynine++;
        }
        System.out.println(eightynine);
        long end = System.currentTimeMillis();
        System.out.println(end - start);
    }
    
    private static int createChain(int n) {
        ArrayList<Integer> chain = new ArrayList<Integer>();
        chain.add(n);
        for(; n != 1 && n != 89;n = squareDigits(n), chain.add(n)){ }
        return chain.get(chain.size() - 1);
    }
    
    private static int squareDigits(int n) {
        if (n < 10) return n * n;
        else return squareDigits(n / 10) + squareDigits(n % 10);
    }
}

File: 05diff/92/squareDigitChains.py
import time 
dictionary = {}
def solution():
    eightynine = 0
    one = 0
    for i in range(1,1000001):
        if createChain(i,dictionary)[-1] == 1:
            one += 1
        else:
            eightynine += 1
    return eightynine

def createChain(num,dictionary):
    chain = [num]
    while num != 1 and num != 89:
        if num in dictionary:
            return [dictionary[num]]
        chain.append(squareDigits(num))
        num = squareDigits(num)
    for i in chain:
        dictionary[i] = chain[-1]
    return chain 
        
def squareDigits(num):
    sum = 0 
    for digit in str(num):
        sum += int(digit) ** 2 
    return sum
  
start = time.time()
print solution()
end = time.time()
print end - start 